pinotdb/db.py

```diff
@@ -53,25 +53,50 @@ def g(self, *args, **kwargs):
     return g
 
 
-def get_description_from_row(row):
-    """
-    Return description from a single row.
 
-    We only return the nameand  type (inferred from the data).
-    """
+def get_description_from_types(types):
     return [
         (
             name,               # name
-            get_type(value),    # type_code
+            type,               # type_code
             None,               # [display_size]
             None,               # [internal_size]
             None,               # [precision]
             None,               # [scale]
             None,               # [null_ok]
         )
-        for name, value in row.items()
+        for name, type in types.items()
```

Comment:
@agrawaldevesh I understand that name/type tuples can be ordered arbitrarily as long as the names match with the results. The problem here is that (in Python < 3.6) there's no guaranteed order for the elements returned by `types.items()`. On each run this function might return a completely different order of columns for the exact same input, and the order of the cursor description will not match the order of the columns in `_results`. It's basically randomizing the order of the description, and only the description.

Reply:
I still think this is right because even in python3, ordering of dict.keys corresponds with dict.items. (As per the references given in https://stackoverflow.com/questions/835092/python-dictionary-are-keys-and-values-always-the-same-order)

Say that the types returned by the new function "get_types_from_rows()" is {"col3": Type.INT, "col2": Type.STRING, "col1": Type.BOOLEAN}. And say the type_keys array is populated as ["col1", "col3", "col2"]. The rows will be returned in the order of type_keys as an array of array:

[ 
[True, 1, "one"],
[False, 5, "five"]
]

And since the get_description_from_types iterates on types.items(), its returned order will match that of type_keys.

Nevertheless, I will fix the code to make it clear that this order in type_keys is the order everywhere. Then the reader won't have to remember this subtlety that the python dict item/keys order matches.

Stay tuned for an update. Thanks for the review !

Reply:
Ah, I see. I wasn't aware of that, thanks for reference! I thought the order could change between calls even if the dict wasn't modified. #TIL :)

But yeah, maybe explicitly using an ordered dict from `collections` or a list of tuples would be clearer.